File: common/internal_model/src/main/scala/weco/catalogue/internal_model/work/WorkData.scala
package weco.catalogue.internal_model.work

import weco.catalogue.internal_model.identifiers.{
  DataState,
  ReferenceNumber,
  SourceIdentifier
}
import weco.catalogue.internal_model.image.ImageData
import weco.catalogue.internal_model.languages.Language
import weco.catalogue.internal_model.locations.Location

/** WorkData should only contain data that we would display on a works page
  * to users.
  */
case class WorkData[State <: DataState](
  title: Option[String] = None,
  otherIdentifiers: List[SourceIdentifier] = Nil,
  // TODO: Move MergeCandidates out of WorkData and into an appropriate part
  // of the WorkState model.  We don't display these to users.
  mergeCandidates: List[MergeCandidate[State#Id]] = Nil,
  alternativeTitles: List[String] = Nil,
  format: Option[Format] = None,
  description: Option[String] = None,
  physicalDescription: Option[String] = None,
  lettering: Option[String] = None,
  createdDate: Option[Period[State#MaybeId]] = None,
  subjects: List[Subject[State#MaybeId]] = Nil,
  genres: List[Genre[State#MaybeId]] = Nil,
  contributors: List[Contributor[State#MaybeId]] = Nil,
  thumbnail: Option[Location] = None,
  production: List[ProductionEvent[State#MaybeId]] = Nil,
  languages: List[Language] = Nil,
  edition: Option[String] = None,
  notes: List[Note] = Nil,
  duration: Option[Int] = None,
  items: List[Item[State#MaybeId]] = Nil,
  holdings: List[Holdings] = Nil,
  collectionLabel: Option[String] = None,
  referenceNumber: Option[ReferenceNumber] = None,
  imageData: List[ImageData[State#Id]] = Nil,
  workType: WorkType = WorkType.Standard,
)

File: catalogue_graph/src/models/pipeline/work_data.py
from typing import Literal

from pydantic import Field

from models.pipeline.collection_path import CollectionPath
from models.pipeline.concept import Concept, Contributor, Genre, Subject
from models.pipeline.holdings import Holdings
from models.pipeline.id_label import Format, Language
from models.pipeline.identifier import SourceIdentifier
from models.pipeline.image import ImageData
from models.pipeline.item import Item
from models.pipeline.location import DigitalLocation
from models.pipeline.note import Note
from models.pipeline.production import ProductionEvent
from models.pipeline.serialisable import ElasticsearchModel

WorkType = Literal["Standard", "Series", "Section", "Collection"]


class WorkData(ElasticsearchModel):
    title: str | None = None
    other_identifiers: list[SourceIdentifier] = Field(default_factory=list)
    alternative_titles: list[str] = Field(default_factory=list)
    format: Format | None = None
    description: str | None = None
    physical_description: str | None = None
    lettering: str | None = None
    created_date: Concept | None = None
    subjects: list[Subject] = Field(default_factory=list)
    genres: list[Genre] = Field(default_factory=list)
    contributors: list[Contributor] = Field(default_factory=list)
    thumbnail: DigitalLocation | None = None
    production: list[ProductionEvent] = Field(default_factory=list)
    languages: list[Language] = Field(default_factory=list)
    edition: str | None = None
    notes: list[Note] = Field(default_factory=list)
    duration: int | None = None
    items: list[Item] = Field(default_factory=list)
    holdings: list[Holdings] = Field(default_factory=list)
    collection_path: CollectionPath | None = None
    reference_number: str | None = None
    image_data: list[ImageData] = Field(default_factory=list)
    work_type: WorkType = "Standard"
    current_frequency: str | None = None
    former_frequency: list[str] = Field(default_factory=list)
    designation: list[str] = Field(default_factory=list)
